packages/vscode-extension/src/project/deviceSession.ts

```diff
@@ -54,26 +49,40 @@ export class DeviceSession implements Disposable {
         })
       : Promise.resolve();
 
+    progressCallback(StartupMessage.Launching);
+    await this.device.launchApp(this.buildResult, this.metro.port, this.devtools.port);
+
+    Logger.debug("Will wait for app ready and for preview");
+    progressCallback(StartupMessage.WaitingForAppToLoad);
+    await Promise.all([waitForAppReady, this.device.startPreview()]);
+    Logger.debug("App and preview ready, moving on...");
+
+    progressCallback(StartupMessage.AttachingDebugger);
+    await this.startDebugger();
+  }
+
+  async restart(progressCallback: ProgressCallback) {
+    return this.launch(progressCallback);
+  }
+
+  async start(
+    deviceSettings: DeviceSettings,
+    previewReadyCallback: PreviewReadyCallback,
+    progressCallback: ProgressCallback
+  ) {
     progressCallback(StartupMessage.BootingDevice);
     await this.device.bootDevice();
     await this.device.changeSettings(deviceSettings);
     progressCallback(StartupMessage.Building);
     this.buildResult = await this.disposableBuild.build;
     progressCallback(StartupMessage.Installing);
     await this.device.installApp(this.buildResult, false);
-    progressCallback(StartupMessage.Launching);
-    await this.device.launchApp(this.buildResult, this.metro.port, this.devtools.port);
 
-    const waitForPreview = this.device.startPreview().then(() => {
+    this.device.startPreview().then(() => {
```

Comment:
NIt: `await` here.

Reply:
awaiting here would block calling this.launch thats below. So this should stay as is